### PR TITLE
feat: CryptoBot payments

### PR DESCRIPTION
**Pull request by gru2007**
I’m just adding new optional payment method

what is it? - https://help.crypt.bot/crypto-pay-api

### CODE REVIEW

**Comment by MrWeez** [comment]
Hi! It looks very cool, but could you change the Russian localization to English?
Thank you for helping to improve cpgg ❤️

**Comment by gru2007** [comment]
Only one thing, for some reason money doesn’t add after sending event with payment. So I added this construction:
```php
try {
            $user->increment('credits', $payment->amount);
        } catch (Exception $exception) {
            throw $exception;
        }
```

I’m sure, that it is not right way to do that. I don’t know what happens in your code.

**Comment by gru2007** [comment]
> Hi! It looks very cool, but could you change the Russian localization to English? Thank you for helping to improve cpgg ❤️

Oh sorry, I have missed that. Couple minutes.

**Comment by 1day2die** [comment]
awesome!

**Comment by 1day2die** [comment]
> Only one thing, for some reason money doesn’t add after sending event with payment. So I added this construction:
> 
> ```
> try {
>             $user->increment('credits', $payment->amount);
>         } catch (Exception $exception) {
>             throw $exception;
>         }
> ```
> 
> I’m sure, that it is not right way to do that. I don’t know what happens in your code.

@IceToast  mind checking this?

**Comment by IceToast** [comment]
So i see you've created a extension with all the necessary information. I will sketch out the payment workflow:

_User payment request -> URL fetching from extension -> Redirecting to payment URL -> ... -> Success redirection to success route of extension -> done_

_Gateway Webhook -> request to extension webhook route -> sets a payment as "paid" -> emits a user payment event -> payment listener updates user credits and possibly any new limits -> done_

The first thing i noticed is that you might want to validate that the webhook request actually wants to update a payment to the PAID status. In your PR it will set it to PAID under any circumstances? Maybe im missing something about the gateway.

For your problem: You should debug into [CryptoBotExtension.php:110](https://github.com/Ctrlpanel-gg/panel/pull/940/commits/69682665c5b587c673c10a341cfcbf839e377a42#diff-0712e40d30808236896ff485db3e8fd82a277dd9fe7119e10479d6a43ff0593bR110) to see whether the event really gets triggered and what da parameters are. If everything is processed correctly the event will be listened to in here: [Listeners/UserPayment.php](https://github.com/Ctrlpanel-gg/panel/blob/development/app/Listeners/UserPayment.php)

**Comment by gru2007** [comment]
@IceToast Thanks, I will come back with result soon.

**Review by IceToast — CHANGES_REQUESTED** [review]

**Comment by MrWeez** [comment]
@gru2007 Is there any news about fixes?

**Comment by gru2007** [comment]
> @gru2007 Is there any news about fixes?

Excuse me for afk, I just haven’t enough time to solve that right now. I’ll do it in near future.

**Comment by MrWeez** [comment]
> > @gru2007 Is there any news about fixes?
> 
> Excuse me for afk, I just haven’t enough time to solve that right now. I’ll do it in near future.

Ok, no problem

**Comment by S0ly** [comment]
updated the branch @gru2007 any news ? :)

File: app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+use Illuminate\Http\JsonResponse;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Redirect;
+use Illuminate\Support\Facades\Log;
+use Illuminate\Support\Facades\Http;
+use App\Notifications\ConfirmPaymentNotification;
+
+/**
+ * Summary of PayPalExtension
```
Comment (MrWeez) [inline app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php:26]: Change or remove this comment

File: app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        $settings = new CryptoBotSettings();
+        try {
+            $response = Http::withHeaders([
+                'Accept' => 'application/json',
+                'Content-Type' => 'application/json',
+                'Crypto-Pay-API-Token' => $settings->api_key
+            ])->post($url, [
+                'amount' => $totalPriceString,
+                'payload' => strval($payment->id),
+                'description' => "Заказ #{$payment->id} - " . $shopProduct->name,
```
Comment (MrWeez) [inline app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php:54]: Replace **Заказ** with **Order**
```suggestion
                'description' => "Order #{$payment->id} - " . $shopProduct->name,
```

File: app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                'currency_type' => 'fiat',
+                'fiat' => $shopProduct->currency_code,
+                'hidden_message' => $settings->hidden_message,
+                'paid_btn_name' => 'callback',
+                'paid_btn_url' => route('payment.CryptoBotSuccess').'?payment=YGqvOf4I'
```
Comment (MrWeez) [inline app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php:59]: I'm not quite sure if it's correct, can you double-check this part of the code? To be more precise, this -> `?payment=YGqvOf4I`

File: app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    {
+        Redirect::route('home')->with('success', 'Your payment is processing')->send();
+        return;
+    }
+
+    static function webhook(Request $request): JsonResponse
+    {
+
+        $settings = new CryptoBotSettings();
+        /***    Проверка подписи     ***/
```
Comment (MrWeez) [inline app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php:83]: Also translate this comment into English
```suggestion
        /***    Signature verification    ***/
```

File: app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        $sign_header = $request->header('crypto-pay-api-signature');
+
+        $secret_key = hash('sha256', $settings->api_key, true);
+        $calculated_signature = hash_hmac('sha256', $request->getContent(), $secret_key);
+
+        if (!hash_equals($sign_header, $calculated_signature)) {
+            return response()->json(['status' => 'invalid sign.']);
+        } 
+        /***                         ***/
```
Comment (MrWeez) [inline app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php:92]: Don't forget to delete unused lines of code (Even comments)
```suggestion
```

File: app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+        $payment = Payment::findOrFail($request->json('payload.payload'));
+        $shopProduct = ShopProduct::findOrFail($payment->shop_item_product_id);
+        $user = User::findOrFail($payment->user_id);
+
+        //update payment
```
Comment (MrWeez) [inline app/Extensions/PaymentGateways/CryptoBot/CryptoBotExtension.php:98]: Keep the style of writing the code, if you used a different type of comment formatting above, then save it to the end of the file
```suggestion
        /***    Update payment    ***/
```

**Review by MrWeez — CHANGES_REQUESTED** [review]

**Comment by S0ly** [comment]
damn good review

**Comment by MrWeez** [comment]
> damn good review

I decided to double-check this PR so that there would be no problems in the future

**Comment by S0ly** [comment]
any update ?

**Comment by S0ly** [comment]
if there is no update soon this PR will be closed

**Comment by S0ly** [comment]
closed for inactivity

**Pull request closed without merging**